### DISCUSSION TITLE
Using NPM modules in Rust web frameworks

### DISCUSSION THREAD

**Discussion by aschrijver** [wasm-bindgen/wasm-bindgen#3488] [Q&A]
In the Leptos discussion section I posted [Combining with other web frameworks / ecosystems](https://github.com/leptos-rs/leptos/discussions/1206), because I want to know the options of integrating existing NPM modules into the Rust-based WebAssembly UI frameworks (Leptos, Yew, etc). In particular [React Flow]() is a component I'd like to integrate with Leptos. It has a large API surface.

Digging a bit into the available tools brought me to wasm-bindgen. However, I have the feeling that there aren't straightforward, easy ways to get the integration going. It looks like hand-coding the JS function imports using `wasm-bindgen` is the way to go.

I am not certain about this though, and most documentation I find here and on the web doesn't give enough clarity. Also I am wondering what future support is expected to wrap components from the rich NPM ecosystem for use in Rust web frameworks (most info is focused on the other way round: use Rust wasm modules in JS-based stacks).

**Comment by daxpedda** [comment]
> It looks like hand-coding the JS function imports using `wasm-bindgen` is the way to go.

That's what `wasm-bindgen` offers indeed.

Generating Rust bindings from JS/TS components isn't currently in scope as far as I'm aware.

**Reply by aschrijver** [reply]
Thank you. Navigating the WebAssembly landscape feels a bit like walking a maze :)
Easing creation of those bindings would open a humongous ecosystem of high quality components.

**Reply by aschrijver** [reply]
WasmEdge has [broad support for Javascript apps](https://wasmedge.org/docs/category/develop-wasm-apps-in-javascript), so I formulated a question if they might support my case:

- https://github.com/WasmEdge/WasmEdge/discussions/2607
